### PR TITLE
Add block deadline to settle payload

### PR DESCRIPTION
**Pull request by Mateo-mro**
# Description
Add block deadline to settle payload. As it is right now, the driver is not aware of the last block number in which they can still include their TX. This information is very important to communicate to drivers, especially in collocation mode.

Feature requested by @harisang .

# Changes
Extend the `/settle` payload (from autopilot) to include the submission deadline in blocks.

## How to test
1. Acceptance tests

### CODE REVIEW

File: crates/autopilot/src/infra/solvers/dto/settle.rs [hunk 1]
```diff
@@ -11,6 +11,10 @@ pub struct Request {
     /// Unique ID of the solution (per driver competition), to settle.
     #[serde_as(as = "serde_with::DisplayFromStr")]
     pub solution_id: u64,
+    /// The maximum number of blocks the autopilot waits for a settlement to
+    /// appear on chain.
+    #[serde_as(as = "serde_with::DisplayFromStr")]
```
Comment (fleupold) [inline crates/autopilot/src/infra/solvers/dto/settle.rs:16]: why is this needed? Do you want to encode this as a string?

**Review by fleupold — CHANGES_REQUESTED** [review]
I think rather than telling the solver how many blocks they have (which can be ambiguous if sent right around the time a new block was mined), we should tell them the last block number that they can still include their tx in.

File: crates/autopilot/src/infra/solvers/dto/settle.rs (hunk 1 above)
Comment (Mateo-mro) [reply crates/autopilot/src/infra/solvers/dto/settle.rs:16]: @fleupold this is a quick draft I pushed just to see how many tests would get affected, it is not the real implementation at all :sweat_smile:

File: crates/autopilot/src/infra/solvers/dto/settle.rs [hunk 2]
```diff
@@ -11,6 +11,8 @@ pub struct Request {
     /// Unique ID of the solution (per driver competition), to settle.
     #[serde_as(as = "serde_with::DisplayFromStr")]
```
Comment (Mateo-mro) [inline crates/autopilot/src/infra/solvers/dto/settle.rs:12]: why do we parse this as a `String`? should I also parse `block_deadline` as a String? :thinking:

File: crates/driver/src/infra/api/routes/settle/dto/solution.rs [hunk 3]
```diff
@@ -8,4 +8,7 @@ pub struct Solution {
     /// Unique ID of the solution (per driver competition), to settle.
     #[serde_as(as = "serde_with::DisplayFromStr")]
     solution_id: u64,
+    #[allow(dead_code)]
+    /// The last block number in which the solution TX can be included
+    pub block_deadline: u64,
```
Comment (Mateo-mro) [inline crates/driver/src/infra/api/routes/settle/dto/solution.rs:13]: I assume we do not need to act on this on our driver (the same as with the `solution_id`.

File: crates/autopilot/src/infra/solvers/dto/settle.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/solvers/dto/settle.rs:12]: Isn't `DisplayFromStr` used everywhere because the value is surrounded by double quotes?

**Comment by squadgazzz** [comment]
The description does not explain why we need this field and how it is expected to be used.

**Comment by Mateo-mro** [comment]
> The description does not explain why we need this field and how it is expected to be used.

@squadgazzz you are right! Thanks! Extending the description!

File: crates/autopilot/src/infra/solvers/dto/settle.rs [hunk 4]
```diff
@@ -11,6 +11,9 @@ pub struct Request {
...
     pub solution_id: u64,
+    /// The last block number in which the solution TX can be included
+    #[serde_as(as = "serde_with::DisplayFromStr")]
+    pub block_deadline: u64,
```
Comment (MartinquaXD) [inline crates/autopilot/src/infra/solvers/dto/settle.rs:16]: Seems more descriptive.
```suggestion
    pub submission_deadline: u64,
```

File: crates/driver/openapi.yml [hunk 5]
```diff
@@ -394,6 +394,10 @@ components:
           description: Id of the solution that should be executed.
           type: string
           example: "123"
+        blockDeadline:
+          description: The last block number in which the solution TX can be included.
+          type: string
+          example: "123"
```
Comment (MartinquaXD) [inline crates/driver/openapi.yml:400]: This can be a number.

File: crates/driver/src/infra/api/routes/settle/dto/solution.rs [hunk 6]
```diff
@@ -8,4 +8,8 @@ pub struct Solution {
     /// Unique ID of the solution (per driver competition), to settle.
     #[serde_as(as = "serde_with::DisplayFromStr")]
     solution_id: u64,
+    #[allow(dead_code)]
+    /// The last block number in which the solution TX can be included
+    #[serde_as(as = "serde_with::DisplayFromStr")]
+    pub block_deadline: u64,
```
Comment (MartinquaXD) [inline crates/driver/src/infra/api/routes/settle/dto/solution.rs:14]:
```suggestion
    pub submission_deadline: u64,
```

File: crates/driver/src/infra/api/routes/settle/dto/solution.rs (hunk 3 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/api/routes/settle/dto/solution.rs:14]: The driver already has a submission deadline in a config file. It should now be read from the request and the config argument should be dropped.

File: crates/driver/openapi.yml (hunk 5 above)
Comment (Mateo-mro) [reply crates/driver/openapi.yml:400]: @MartinquaXD I agree, it was a number but I changed it to a `String`, because `solutionId` could be a number but it is sent a `String` , so I thought "there has to be a good reason for it", plus, to keep consistency. What is the reason `solutionId` is parsed as a `String` even tho it is an `u64`? should I keep the consistency?

File: crates/driver/src/infra/config/file/mod.rs [hunk 7]
```diff
@@ -167,10 +162,6 @@ fn default_retry_interval() -> Duration {
     Duration::from_secs(2)
 }
 
-fn default_max_confirm_time() -> Duration {
-    Duration::from_secs(120)
```
Comment (Mateo-mro) [inline crates/driver/src/infra/config/file/mod.rs:171]: I am surprised this was set to 2 minutes?

File: crates/driver/openapi.yml (hunk 5 above)
Comment (MartinquaXD) [reply crates/driver/openapi.yml:400]: `SolutionId` is just some sort of identifier. Currently it's always a number but there is nothing preventing us to turn it into a string if that seems more helpful. Making it a string from the start means if we ever choose to change the underlying format no other code would have to be changed.
The block number on the other hand is and will always be a number so there is no reason to be extra flexible there.

File: crates/driver/src/infra/config/file/mod.rs (hunk 7 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/config/file/mod.rs:171]: Yeah, the defaults in the driver are still a bit rough. Although in practice we'll likely never hit the full 2 minutes because the proposed solution would start to revert much earlier which also leads the driver to abort the submission.

File: crates/autopilot/src/infra/solvers/dto/settle.rs [hunk 8]
```diff
@@ -11,6 +11,8 @@ pub struct Request {
     /// Unique ID of the solution (per driver competition), to settle.
     #[serde_as(as = "serde_with::DisplayFromStr")]
     pub solution_id: u64,
+    /// The last block number in which the solution TX can be included
+    pub submission_deadline: u64,
```
Comment (fleupold) [inline crates/autopilot/src/infra/solvers/dto/settle.rs:15]: nit: might be nice to make the fact that this is a block number (and not e.g. a unix timestamp) more obvious by including it in the name (something like `latest_block`) and avoiding people having to read the openapi docs?

File: crates/driver/src/domain/mempools.rs [hunk 9]
```diff
@@ -104,11 +105,15 @@ impl Mempools {
         let result = async {
             loop {
                 // Wait for the next block to be mined or we time out.
-                if tokio::time::timeout_at(deadline, block_stream.next())
-                    .await
-                    .is_err()
-                {
-                    tracing::info!(?hash, "tx not confirmed in time, cancelling");
+                let current_block = self.ethereum.current_block().borrow().number;
+                if current_block > submission_deadline {
```
Comment (squadgazzz) [inline crates/driver/src/domain/mempools.rs:109]: It looks like it doesn't make sense to check this condition more often than a new block is mined. Since there is access to `block_stream` already, it can be used instead of loop like
```
while let Some(block) = block_stream.next().await {
...
}
```

File: crates/driver/openapi.yml (hunk 5 above)
Comment (MartinquaXD) [reply crates/driver/openapi.yml:400]: Why was this resolved? It's still a sting in the openapi spec.

File: crates/driver/src/domain/mempools.rs [hunk 10]
```diff
@@ -102,13 +103,16 @@ impl Mempools {
...
         let result = async {
-            loop {
+            while let Some(block) = block_stream.next().await {
                 // Wait for the next block to be mined or we time out.
-                if tokio::time::timeout_at(deadline, block_stream.next())
-                    .await
-                    .is_err()
-                {
-                    tracing::info!(?hash, "tx not confirmed in time, cancelling");
+                if block.number > submission_deadline {
```
Comment (MartinquaXD) [inline crates/driver/src/domain/mempools.rs:108]: I think at the moment this cancels the tx one block later than it could.
Because the deadline gets checked before checking the blockchain if our tx was mined this check currently has to happen 1 block after the deadline. If this check were to move into the `TxStatus::Pending` case of the `match` below we could already abort if `block.number == submission_deadline`.

File: crates/driver/src/domain/mempools.rs [hunk 11]
```diff
@@ -143,6 +147,7 @@ impl Mempools {
                     }
                 }
             }
+            panic!("Block stream finished unexpectedly");
```
Comment (MartinquaXD) [inline crates/driver/src/domain/mempools.rs:150]: This should only be an `error` log IMO. We'll learn about that via alerts as well but it doesn't crash the pod unnecessarily.

File: crates/driver/src/infra/api/routes/settle/mod.rs [hunk 12]
```diff
@@ -14,13 +14,15 @@ pub(in crate::infra::api) fn settle(router: axum::Router<State>) -> axum::Router
...
-    _: axum::Json<dto::Solution>,
+    solution: axum::Json<dto::Solution>,
 ) -> Result<(), (hyper::StatusCode, axum::Json<Error>)> {
     let competition = state.competition();
     let auction_id = competition.auction_id().map(|id| id.0);
     let handle_request = async {
         observe::settling();
-        let result = competition.settle().await;
+        let result = competition
+            .settle(solution.submission_deadline_latest_block)
```
Comment (MartinquaXD) [inline crates/driver/src/infra/api/routes/settle/mod.rs:24]: Seems a bit odd to have the submission deadline on the solution instead of the competition. 🤔 
I guess the reason is that the competition gets created before the `/settle` call happens which is where we learn about the deadline?
Ideally this would get changed because this conceptually allows the existence of 2 solutions for the same auction with different deadlines.

File: crates/driver/src/tests/setup/driver.rs [hunk 13]
```diff
@@ -135,6 +135,7 @@ pub fn reveal_req() -> serde_json::Value {
 pub fn settle_req() -> serde_json::Value {
     json!({
         "solutionId": "0",
+        "submissionDeadlineLatestBlock": 0,
```
Comment (MartinquaXD) [inline crates/driver/src/tests/setup/driver.rs:138]: Does this not get used? A submission deadline of block 0 should never work.

File: crates/driver/openapi.yml [hunk 14]
```diff
@@ -28,7 +28,7 @@ paths:
             - `sell`: amount is in sell_token, out_amount is in buy_token
           schema:
             type: string
-            enum: ["buy", "sell"]
+            enum: [ "buy", "sell" ]
```
Comment (MartinquaXD) [inline crates/driver/openapi.yml:31]: Out of curiosity how did this get formatted? A bit unfortunate to have these changes in unrelated PRs. Maybe we can have a CI step that verifies that this is formatted correctly. 🤔

File: crates/driver/src/infra/api/routes/settle/mod.rs (hunk 12 above)
Comment (Mateo-mro) [reply crates/driver/src/infra/api/routes/settle/mod.rs:24]: @MartinquaXD the deadline is on `/settle` payload meaning that the autopilot is telling the driver how much time it has to settle it. I believe this was the main point from Haris. What is your suggestion here? :thinking:

File: crates/driver/openapi.yml (hunk 14 above)
Comment (Mateo-mro) [reply crates/driver/openapi.yml:31]: @MartinquaXD sorry, it is my IDE which does this without my permission :sweat_smile: reverting it.

File: crates/driver/openapi.yml (hunk 5 above)
Comment (Mateo-mro) [reply crates/driver/openapi.yml]: Ups, overwritten while fighting with my IDE for its automatic changes, fixed, good catch!

File: crates/driver/src/domain/mempools.rs (hunk 11 above)
Comment (squadgazzz) [reply crates/driver/src/domain/mempools.rs:149]: In my previous comment, I was thinking about canceling the settlement and returning a meaningful error. Panic doesn't make much sense here IMO. I think we will get alerted if all the settlements fail for some period of time.

File: crates/driver/src/infra/api/routes/settle/mod.rs (hunk 12 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/api/routes/settle/mod.rs:24]: Misread the code because the `solution` variable name was a bit misleading. This actually the `/settle` requst. So the value is where it should be the naming of the variable and struct is just a bit confusing.

File: crates/driver/src/domain/mempools.rs (hunk 11 above)
Comment (Mateo-mro) [reply crates/driver/src/domain/mempools.rs]: Thanks guys for the discussion! I removed the panic :ok_hand:

File: crates/autopilot/src/infra/solvers/dto/settle.rs [hunk 15]
```diff
@@ -11,6 +11,8 @@ pub struct Request {
     /// Unique ID of the solution (per driver competition), to settle.
     #[serde_as(as = "serde_with::DisplayFromStr")]
     pub solution_id: u64,
+    /// The last block number in which the solution TX can be included
+    pub submission_deadline_latest_block: u64,
```
Comment (squadgazzz) [inline crates/autopilot/src/infra/solvers/dto/settle.rs:15]: nit: maybe something a bit shorter. Also, in different places, this param is still named `submission_deadline`.
```suggestion
    pub deadline_block_num: u64,
```

Comment (MartinquaXD) [reply crates/autopilot/src/infra/solvers/dto/settle.rs:15]: 3 different naming nits from 3 people on the same variable name. 😂

Comment (squadgazzz) [reply crates/autopilot/src/infra/solvers/dto/settle.rs:15]: I didn't notice it in the resolved comments lol

Comment (Mateo-mro) [reply crates/autopilot/src/infra/solvers/dto/settle.rs:15]: Ok guys @fleupold  @MartinquaXD @squadgazzz , let's discuss here the name, so we all agree on one :sweat_smile: 

So far the proposals are:

1) `submission_deadline_latest_block`
2) `deadline_block_num`
3) `deadline_block`

IMO, I think the shorter version name proposed is not as clear as the current one. I believe the current one (1) is the best of all of them.

Comment (MartinquaXD) [reply crates/autopilot/src/infra/solvers/dto/settle.rs:15]: I'm fine with `1` since the slightly longer name is not used a lot anyway.

File: crates/driver/src/domain/mempools.rs [hunk 16]
```diff
@@ -41,12 +41,13 @@ impl Mempools {
         &self,
         solver: &Solver,
         settlement: &Settlement,
+        submission_deadline: u64,
```
Comment (fleupold) [inline crates/driver/src/domain/mempools.rs:44]: nit: add `type BlockNo = u64;` to avoid confusion of what this u64 is here as well (could in theory also be a timestamp)

File: crates/driver/src/domain/mempools.rs [hunk 17]
```diff
@@ -126,6 +117,17 @@ impl Mempools {
                     TxStatus::Executed => return Ok(hash.clone()),
                     TxStatus::Reverted => return Err(Error::Revert(hash.clone())),
                     TxStatus::Pending => {
+                        // Check if the current block reached the submission deadline block number
+                        if block.number == submission_deadline {
```
Comment (fleupold) [inline crates/driver/src/domain/mempools.rs:121]: In arbitrum we may not see every block (and in the spirit of defensive programming), let's use

```suggestion
                        if block.number >= submission_deadline {
```

**Review by fleupold — APPROVED** [review]
Comparison should use `>=` to be safe against reorgs and otherwise skipped blocks.

**Review by squadgazzz — APPROVED** [review]
LGTM

**Review by MartinquaXD — APPROVED** [review]
